src/aggregated/circulating_supply/circulating_supply_provider.rs

```diff
@@ -0,0 +1,193 @@
+        }
+        Ok(Some(supply)) => {
+            info!(
+                target: crate::AGGREGATED,
+                "Circulating supply for the timestamp {} was already computed: {}",
+                block_timestamp,
+                supply
+            );
+            Ok(None)
+        }
+        Err(err) => Err(err),
+    }
+}
+
+async fn compute_circulating_supply_for_block(
+    pool: &Database<PgConnection>,
+    view_client: &Addr<near_client::ViewClientActor>,
+    block: &models::Block,
+) -> Result<CirculatingSupply, String> {
+    let block_timestamp = block
+        .block_timestamp
+        .to_u64()
+        .ok_or("`block_timestamp` expected to be u64")?;
+    let block_height = block
+        .block_height
+        .to_u64()
+        .ok_or("`block_height` expected to be u64")?;
+    let total_supply = u128::from_str_radix(&*block.total_supply.to_string(), 10)
+        .map_err(|_| "`total_supply` expected to be u128")?;
+
+    let lockup_account_ids =
+        accounts::get_lockup_account_ids_at_block_height(&pool, &block_height).await?;
+
+    let mut lockups_locked_tokens: u128 = 0;
+    let mut unfinished_lockup_contracts_count: i32 = 0;
+
+    for lockup_account_id in &lockup_account_ids {
+        let state =
+            lockup::get_lockup_contract_state(&view_client, lockup_account_id, &block_height)
+                .await
+                .map_err(|err| format!("Failed to get lockup contract: {}", err))?;
+        let code_hash =
+            account_details::get_contract_code_hash(&view_client, lockup_account_id, &block_height)
+                .await?;
+        let is_lockup_with_bug = lockup::is_bug_inside_contract(&code_hash, lockup_account_id)?;
+        let locked_amount = state
+            .get_locked_amount(block_timestamp, is_lockup_with_bug)
+            .0;
+        lockups_locked_tokens += locked_amount;
+        if locked_amount > 0 {
+            unfinished_lockup_contracts_count += 1;
+        }
+    }
+
+    // The list is taken from the conversation with Yessin
+    let foundation_locked_account_ids: [near_primitives::types::AccountId; 2] =
+        ["lockup.near".to_string(), "contributors.near".to_string()];
+    let mut foundation_locked_tokens: u128 = 0;
+    for account_id in &foundation_locked_account_ids {
+        foundation_locked_tokens +=
+            account_details::get_account_balance(&view_client, &account_id, &block_height).await?;
+    }
+
+    let circulating_supply: u128 = total_supply - foundation_locked_tokens - lockups_locked_tokens;
+
+    Ok(CirculatingSupply {
+        computed_at_block_timestamp: BigDecimal::from(block_timestamp),
+        computed_at_block_hash: (&block.block_hash).to_string(),
+        circulating_tokens_supply: BigDecimal::from_str(circulating_supply.to_string().as_str())
+            .map_err(|_| "`circulating_tokens_supply` expected to be u128")?,
+        total_tokens_supply: BigDecimal::from_str(total_supply.to_string().as_str())
```

Comment:
This is slightly easier to read:

```suggestion
        total_tokens_supply: BigDecimal::from_str(&total_supply.to_string())
```

Alternatively, you may also go with:

```
        total_tokens_supply: total_supply.to_string().parse()
```